### ISSUE TITLE
Seperate node module for embedding and hosts?

### ISSUE THREAD

**Issue by DonWinston** [honestbleeps/Reddit-Enhancement-Suite#4224]
The code you've written to display a preview per URL is really useful and I haven't seen a library on the internet do it as well as you guys. I'd like to pull some of that code into my own project but feel like it would just make a great npm module, would you guys be okay with a published npm module that credits and links back to RES? 

The module would take a URL and create embeddable HTML, and have a config file for API keys.

**Comment by honestbleeps** [comment]
Actually, I've thought about making it a standalone library before, during the time I was working on a HoverZoom alternative called BetterZoom, and I just never got around to it. I hadn't necessarily thought `npm` specifically, but I can see that being useful.

I'd definitely support this idea. My preference would be that me and/or the RES team "own" the repo and the publishing of the module, but we'd certainly make anyone willing to make active contributions to it a collaborator. 

The catch, however, is that it's a bit more complex than simply consuming a URL and returning embeddable HTML. It might make more sense for it to return metadata that is used to create embeddable HTML.  The HTML generated by RES is rather RES-specific, and what you really want is for an application to receive the direct URL to the file, and/or the metadata required to wrap your own player or iframe around something. That way, it still has control over the style, markup and output.

Thoughts?

**Comment by DonWinston** [comment]
Yeah as I'm diving into code, I'm realizing how much of the controls and stuff is specific to RES. 

Right now I'm thinking of making two parts, one that takes in a url and outputs metadata like you said and the other, a react component that takes in the metadata and actually displays the media. 

This is something I want to do pretty soon, so I may just start doing it on my own. If you guys don't get to it, I can contribute the code or we can collaborate somehow.

**Comment by Ajedi32** [comment]
I've thought about doing this in the past as well. One potential problem with it though is that RES is licensed under GPL, and it's [somewhat questionable](https://samgentle.com/posts/2015-03-31-illegal-npm-modules) as to whether it's legal to use a GPL-licensed node module as a dependency of a non-GPL-licensed node module. This is somewhat problematic for Node-land, where the MIT license tends to be the prefered standard. 

This leaves us with a few options:

1. We could release a GPL-licensed module based on RES's code. Unfortunately, this would mean very few projects would be able to legally use that module without changing their project's license.
2. We could attempt to contact all the authors of the code under `lib/modules/hosts` (as determined by git) and ask them to relicense their contributions under the MIT license. Given the number of people who have contributed modules to RES, that might be difficult
3. We could start from scratch with a MIT-licensed clean-room implementation of the code under `lib/modules/hosts`, and modify RES to depend on that new implementation, encouraging new contributors to contribute new hosts to the library rather than directly to RES.

I'm personally in favor of a combination of 2 and 3. Ask authors to relicense their code where possible, and for those who either refuse or don't respond, reimplement the hosts they worked on via a clean room design.

**Comment by thybag** [comment]
Just to chip in, I'm very happy for any of my contributions to the media hosts to be used under MIT. 

Additionally I like the idea of scoping the library to just parsing urls & providing embed metadata. Keeps it focused + maximises the places it could be useful .
